### PR TITLE
feat(PopoverLayout ModalLayout): new component, also update DialogLayout

### PR DESCRIPTION
**Pull request by laraharrow**
- [ ] ♿️ Accessibility addressed
- [ ] 🌏 Localization & Internationalization addressed
- [ ] 🖼 Image Snapshot coverage
- [ ] 📚 Documentation updated
- [ ] 🧳 Bundle size impact addressed
- [ ] 🏁 Performance impacts addressed
- [ ] 👾 Browsers tested
  - [ ] Chrome
  - [ ] Firefox
  - [ ] Safari
  - [ ] IE11

### CODE REVIEW

**Comment by laraharrow** [comment]
Ok I think this is ready for review!
A few notes here:
1. I made a few changes to the snapshots based on the new code that was added. @lukelooker maybe we can chat about what is the best way to have every option well represented but with no repetition?
2. I don't think that the `DialogFooter` storybook examples are exactly relatable to the use cases we have in helltool, not sure if we want to update that.

**Review by lukelooker — CHANGES_REQUESTED** [review]
Needs rebase :)

File: packages/components/src/Dialog/Layout/DialogHeader/DialogHeader.story.tsx [hunk 1]
```diff
@@ -36,9 +36,7 @@ export const Basic = () => <DialogHeader>Heading</DialogHeader>
 
 export const HideClose = () => <DialogHeader hideClose>Heading</DialogHeader>
 
-export const Detail = () => (
-  <DialogHeader detail="Detail text">Heading</DialogHeader>
-)
```
Comment (lukelooker) [inline packages/components/src/Dialog/Layout/DialogHeader/DialogHeader.story.tsx:41]: Can we revert this change?

File: packages/components/src/Dialog/Layout/DialogHeader/DialogHeader.test.tsx [hunk 2]
```diff
@@ -42,11 +42,6 @@ describe('DialogHeader', () => {
     expect(screen.getByText('Close')).toBeInTheDocument()
   })
 
-  test('Detail text visible', () => {
-    renderWithTheme(<Detail />)
-    expect(screen.getByText('Detail text')).toBeInTheDocument()
-  })
-
```
Comment (lukelooker) [inline packages/components/src/Dialog/Layout/DialogHeader/DialogHeader.test.tsx:49]: Let's revert this change and the one below (effectively just moves around the test and I'd prefer to limit the amount of change since we've had "jitter" on these components recently)

File: packages/components/src/__mocks__/DialogExampleLayout.tsx [hunk 3]
```diff
@@ -36,13 +36,11 @@ export const DialogExampleLayout: FC<{ header: string }> = ({
   return (
     <DialogLayout
       header={header}
-      footer={
```
Comment (laraharrow) [inline packages/components/src/__mocks__/DialogExampleLayout.tsx:39]: @lukelooker this is what I was mentioning about the stories

File: packages/components/src/Dialog/Layout/DialogLayout.story.tsx [hunk 4]
```diff
@@ -25,65 +25,64 @@
...
-export interface DialogLayoutExampleProps {
-  header?: string
-  footer?: string
-  length?: 'short' | 'long'
-}
+const extraButton = (
+  <ButtonTransparent color="neutral" size="medium">
+    Cancel
+  </ButtonTransparent>
+)
```
Comment (lukelooker) [inline packages/components/src/Dialog/Layout/DialogLayout.story.tsx:37]: Can this set of stories be restore to how they were used before so that we can verify that no of the associated image snapshots were altered unexpectedly?

File: packages/components/src/Dialog/Layout/DialogLayout.story.tsx [hunk 5]
```diff
@@ -25,65 +25,64 @@
...
-  ...Basic.args,
-  footer: 'Footer text',
-}
+export const HeaderCloseButton = () => (
+  <Box bg="ui1">
+    <DialogLayout header="Header" headerCloseButton={true} footer="Footer">
+      <ConstitutionShort />
+    </DialogLayout>
+  </Box>
+)
```
Comment (lukelooker) [inline packages/components/src/Dialog/Layout/DialogLayout.story.tsx:61]: This is a net-new test and can be kept.

File: packages/components/src/Dialog/Layout/DialogLayout.story.tsx [hunk 6]
```diff
@@ -25,65 +25,64 @@
...
+      footer={
+        <DialogFooter secondary={extraButton}>
+          <Button>Primary</Button>
+        </DialogFooter>
+      }
+    >
+      <ConstitutionShort />
+    </DialogLayout>
+  </Box>
+)
```
Comment (lukelooker) [inline packages/components/src/Dialog/Layout/DialogLayout.story.tsx:76]: This is a net-new test and can be kept.

File: packages/components/src/Dialog/Layout/DialogLayout.story.tsx [hunk 7]
```diff
@@ -25,65 +25,64 @@
...
 
-const NoPaddingTemplate: Story<DialogLayoutExampleProps> = () => (
+export const NoPadding = () => (
   <Box bg="ui1">
     <DialogContent hasFooter={false} hasHeader={false}>
       <ConstitutionShort />
     </DialogContent>
   </Box>
 )
```
Comment (lukelooker) [inline packages/components/src/Dialog/Layout/DialogLayout.story.tsx:84]: I'm okay with this change.

File: packages/components/src/Dialog/Layout/DialogLayout.story.tsx (hunk 4 above)
Comment (lukelooker) [reply packages/components/src/Dialog/Layout/DialogLayout.story.tsx:37]: Specifically `DialogLayoutExampleProps`, `Template`, `Basic`, `Header` & `Footer`

File: packages/components/src/Modal/ModalFooter/ModalFooter.tsx [hunk 8]
```diff
@@ -47,12 +48,16 @@ const ModalFooterLayout: FC<ModalFooterProps> = ({
...
   ...props
 }) => (
-  <Space as="footer" reverse between {...props}>
-    <Space reverse>{children}</Space>
-    {secondary && <Space>{secondary}</Space>}
-  </Space>
+  <Flex as="footer" justifyContent="flex-end" {...props}>
+    {secondary && <FooterLayout>{secondary}</FooterLayout>}
+    {children}
+  </Flex>
```
Comment (lukelooker) [inline packages/components/src/Modal/ModalFooter/ModalFooter.tsx:54]: This will break the existing `DialogFooter` layout for "secondary" I think – "secondary" is supposed to go on the left-hand side of the `DialogFooter`

```
[secondary]                                                   [children]
```

e.g.:

```
[Add Another]                                                       [ Cancel,  Save Changes ]
```

File: packages/components/src/Modal/ModalHeader/ModalHeader.tsx [hunk 9]
```diff
@@ -85,4 +85,5 @@ export const ModalHeader = styled(ModalHeaderLayout)<ModalHeaderProps>`
   align-items: center;
   display: flex;
   flex-shrink: 0;
+  width: 100%;
```
Comment (lukelooker) [inline packages/components/src/Modal/ModalHeader/ModalHeader.tsx:88]: I'm not remember why we needed to add this (I think it's fine but always great to see a note on the PR noting why a change to a component was needed

File: packages/components/src/Popover/Layout/PopoverFooter/PopoverFooter.story.tsx [hunk 10]
```diff
@@ -25,33 +25,26 @@
...
   title: 'PopoverFooter',
 }
 
-export const Basic = () => <PopoverFooter>Footer Text</PopoverFooter>
-Basic.parameters = {
-  storyshots: {
-    disable: true,
-  },
-}
+export const Basic = () => <PopoverFooter />
```
Comment (lukelooker) [inline packages/components/src/Popover/Layout/PopoverFooter/PopoverFooter.story.tsx:36]: Pretty sure this still needs to have a child `Footer Text` ?

File: packages/components/src/__mocks__/DialogExampleLayout.tsx [hunk 11]
```diff
@@ -36,13 +36,11 @@ export const DialogExampleLayout: FC<{ header: string }> = ({
...
-          <Button onClick={closeModal}>Done Reading</Button>
-          <ButtonTransparent color="neutral" onClick={closeModal}>
-            Finish Later
-          </ButtonTransparent>
-        </>
+      footer={<Button onClick={closeModal}>Done Reading</Button>}
+      footerSecondary={
+        <ButtonTransparent color="neutral" onClick={closeModal}>
+          Finish Later
+        </ButtonTransparent>
```
Comment (lukelooker) [inline packages/components/src/__mocks__/DialogExampleLayout.tsx:43]: This change needs to be reversed as it alters the result of a lot of image snapshots. The previous version was the expected design (no use of `footerSecondary` in these)

**Review by lukelooker — CHANGES_REQUESTED** [review]
Haven't dug into the `Popover` specific changes as of yet - I'd like to start with just having `Dialog*` snapshots stay identical (no changes). Additional snapshots is fine but I'd expect to see no changes to the existing "known good" snapshots.

File: packages/components/src/Dialog/Layout/DialogHeader/DialogHeader.story.tsx (hunk 1 above)
Comment (laraharrow) [reply packages/components/src/Dialog/Layout/DialogHeader/DialogHeader.story.tsx:41]: I updated this change but as of right now its set as `storyshots: { disable: true },` so it shouldn't affect snapshots.

File: packages/components/src/Modal/ModalFooter/ModalFooter.tsx (hunk 8 above)
Comment (laraharrow) [reply packages/components/src/Modal/ModalFooter/ModalFooter.tsx:54]: That's one of the problems with my understanding of the `figma` the was that the component is displayed there there is no example with `secondary` so I assumed the `cancel` word there was the `secondary` prop. 
https://www.figma.com/file/zd6MjUXIIstvG7gTJWaX2Y/Looker-Components?node-id=3%3A569
So I actually change the code to accommodate that🤦🏽‍♀️.

File: packages/components/src/Modal/ModalHeader/ModalHeader.tsx (hunk 9 above)
Comment (laraharrow) [reply packages/components/src/Modal/ModalHeader/ModalHeader.tsx:88]: My bed, when we chat last there was a bug I was trying to fix. You had suggested adding that would maybe solve the problem. I added thinking I would go back and check and after solving the bug I forgot it that was there.

File: packages/components/src/Popover/Layout/PopoverFooter/PopoverFooter.story.tsx (hunk 10 above)
Comment (laraharrow) [reply packages/components/src/Popover/Layout/PopoverFooter/PopoverFooter.story.tsx:36]: Because when nothing is passed we would still have the `Done` available there is a chance that nothing needs to be passed. Am I missing something here?

Comment (lukelooker) [reply packages/components/src/Popover/Layout/PopoverFooter/PopoverFooter.story.tsx:36]: Ah! My bad - that makes sense.

File: packages/components/src/Dialog/Layout/DialogLayout.story.tsx [hunk 12]
```diff
@@ -74,16 +68,51 @@ Full.args = {
...
-const NoPaddingTemplate: Story<DialogLayoutExampleProps> = () => (
+const extraButton = (
+  <ButtonTransparent color="neutral" size="medium">
+    Cancel
+  </ButtonTransparent>
+)
+
+export const HeaderDetail = () => (
+  <Box bg="ui1">
+    <DialogLayout header="Header" headerDetail={extraButton} footer="Footer">
```
Comment (lukelooker) [inline packages/components/src/Dialog/Layout/DialogLayout.story.tsx:79]:
```suggestion
    <DialogLayout header="Header" headerDetail="Detail text" footer="Footer">
```

File: packages/components/src/Dialog/Layout/DialogLayout.story.tsx [hunk 13]
```diff
@@ -74,16 +68,51 @@ Full.args = {
   ...Header.args,
 }
 
-const NoPaddingTemplate: Story<DialogLayoutExampleProps> = () => (
+const extraButton = (
+  <ButtonTransparent color="neutral" size="medium">
+    Cancel
+  </ButtonTransparent>
+)
```
Comment (lukelooker) [inline packages/components/src/Dialog/Layout/DialogLayout.story.tsx:75]:
```suggestion
```

File: packages/components/src/Dialog/Layout/DialogLayout.story.tsx [hunk 14]
```diff
@@ -74,16 +68,51 @@ Full.args = {
...
+
+export const HeaderDetail = () => (
+  <Box bg="ui1">
```
Comment (lukelooker) [inline packages/components/src/Dialog/Layout/DialogLayout.story.tsx:78]: I worry that folks previewing these components might come to the false assumption that they have a `ui1` background if they don't look at the source.

So i'd prefer not to wrap see these in a box with background color since it's not something we expect to see in real implementations.

File: packages/components/src/Dialog/Layout/DialogLayout.story.tsx [hunk 15]
```diff
@@ -74,16 +68,51 @@ Full.args = {
...
+
+export const FooterSecondary = () => (
+  <Box bg="ui1">
+    <DialogLayout
+      header="Header"
+      footer={
+        <DialogFooter secondary={extraButton}>
+          <Button>Primary</Button>
+        </DialogFooter>
+      }
```
Comment (lukelooker) [inline packages/components/src/Dialog/Layout/DialogLayout.story.tsx:101]: `DialogLayout` already wraps content output in `footer` within another `DialogFooter` so it's not needed here. Should instead be:

```suggestion
      footer={<Button>Primary</Button>
      footerSecondary="Secondary text"
```

File: packages/components/src/Dialog/Layout/DialogLayout.story.tsx [hunk 16]
```diff
@@ -74,16 +68,51 @@ Full.args = {
...
+  <Box bg="ui1">
+    <DialogLayout header="Header" headerDetail={extraButton} footer="Footer">
+      <ConstitutionShort />
+    </DialogLayout>
+  </Box>
+)
+
+export const HeaderCloseButton = () => (
+  <Box bg="ui1">
+    <DialogLayout header="Header" headerCloseButton={true} footer="Footer">
```
Comment (lukelooker) [inline packages/components/src/Dialog/Layout/DialogLayout.story.tsx:87]:
```suggestion
    <DialogLayout header="Header" headerCloseButton footer="Footer">
```

File: packages/components/src/Dialog/Layout/DialogLayout.story.tsx (hunk 14 above)
Comment (lukelooker) [reply packages/components/src/Dialog/Layout/DialogLayout.story.tsx:78]: NOTE: Please remove from all of these stories for consistency unless you see a specific benefit to keeping the background color (let me know if so - happy to discuss)

File: packages/components/src/Dialog/Layout/DialogLayout.story.tsx [hunk 17]
```diff
@@ -26,45 +26,39 @@
...
     </DialogLayout>
   </Box>
 )
 
-export const Basic = Template.bind({})
-Basic.args = {
-  length: 'short',
-}
+export const Content = Template.bind({})
+Content.args = {}
```
Comment (lukelooker) [inline packages/components/src/Dialog/Layout/DialogLayout.story.tsx:51]: Any specific reason to rename this story? Most of our `.story` files start with a story named "Basic" so it'd be nice to keep with that convention unless there's a specific reason not to do so.

Also the `length` option made it easier to compose the stories below - wondering why you choose to remove that functionality and hard-code it instead?

```suggestion
export const Basic = Template.bind({})
Basic.args = {
  length: 'short'
}
```

File: packages/components/src/Dialog/Layout/DialogLayout.story.tsx [hunk 18]
```diff
@@ -74,16 +68,51 @@ Full.args = {
...
 }
 
-const NoPaddingTemplate: Story<DialogLayoutExampleProps> = () => (
+const extraButton = (
+  <ButtonTransparent color="neutral" size="medium">
+    Cancel
+  </ButtonTransparent>
+)
+
+export const HeaderDetail = () => (
```
Comment (lukelooker) [inline packages/components/src/Dialog/Layout/DialogLayout.story.tsx:77]: This story could leverage the standard `Template` above by being rewritten as:

```suggestion
export const HeaderDetail = Template.bind({})
HeaderDetail.args = {
  ...Content.args,
  header: 'Header',
  headerDetail: 'Header detail text',
}
```

If you need to leverage the complete story in a Jest test you can do so by following this article: https://storybook.js.org/docs/react/workflows/unit-testing

TLDR: 

```
it('renders the button in the primary state', () => {
  render(<HeaderDetail {...HeaderDetail.args} />);
  expect(screen.getByText('Header detail text');
});
```

Note the `{...HeaderDetail.args} ` that spreads the args onto the component is a key piece here.

File: packages/components/src/__mocks__/DialogExampleLayout.tsx (hunk 3 above)
Comment (lukelooker) [reply packages/components/src/__mocks__/DialogExampleLayout.tsx:39]: This story should be restored to the way it was before - the intention (in this example) is to have the two buttons both on the left-side and not to have anything in the  `secondary`  (left) position.

**Review by lukelooker — CHANGES_REQUESTED** [review]
Hey Lara - 

With the work to prepare 2.0 to launch we found quite a few spots where we had regressions in Dialog layouts (as well as the existing `PopoverContent` behavior so I'd like to be a bit more cautious moving forward so we don't accidentally cause any new issues. 

With that in mind I'd like to break this PR into two seperate PRs:

1. refactor: Extract new `ModalLayout` from `DialogLayout`
2. feat: Add new `PopoverLayout` component

Ideally the initial PR would be able to leave all existing stories and tests for `DialogLayout` intact (adding new ones is totally fine of course). 

I've made a bunch of additional comments as well as unresolving some unaddressed comments from my previous review that should help get the first PR ready to go (the "refactor" portion). 

If you need help with doing the git-magic to pull this apart into a couple of separate PRs I'd be happy to jump on a call with you and walk you through doing so.

File: packages/components/src/Dialog/Layout/DialogLayout.story.tsx (hunk 17 above)
Comment (laraharrow) [reply packages/components/src/Dialog/Layout/DialogLayout.story.tsx:51]: the length option is not used anywhere as `long` so I didn't see the point of having it.

File: packages/components/src/Dialog/Layout/DialogLayout.story.tsx (hunk 14 above)
Comment (laraharrow) [reply packages/components/src/Dialog/Layout/DialogLayout.story.tsx:78]: yeah I totally see that I was going off a different example that was already there that had that in place. I'm happy to remove it.

File: packages/components/src/Dialog/Layout/DialogLayout.story.tsx (hunk 15 above)
Comment (laraharrow) [reply packages/components/src/Dialog/Layout/DialogLayout.story.tsx:101]: oops yep, that totally make sense, my bad!

File: packages/components/src/Dialog/Layout/DialogHeader/DialogHeader.test.tsx (hunk 2 above)
Comment (laraharrow) [reply packages/components/src/Dialog/Layout/DialogHeader/DialogHeader.test.tsx:49]: I think this is done!

File: packages/components/src/Dialog/Layout/DialogLayout.story.tsx (hunk 18 above)
Comment (laraharrow) [reply packages/components/src/Dialog/Layout/DialogLayout.story.tsx:77]: Why should we do this change?
